### DISCUSSION TITLE
mosquitto_passwd

### DISCUSSION THREAD

**Discussion by tobit6795** [Freetz-NG/freetz-ng#463] [Allgemein] [answered]
i noticed after installing freetz with mosquitto broker that the mosquitto_passwd utility is not available.
so first question, can we add mosquitto_passwd?
second question, is user:pass authentication even possible at the moment?

**Comment by fda77** [comment]
Feel free to add it

**Comment by pfichtner** [comment]
@tobit6795 How useful is it to have mosquitto_passwd present as binary on the router? Really needed?

**Reply by tobit6795** [reply]
well, i wouldnt say "really needed", but i like the idea of providing / configuring user:password over the WebUI. to make this possible i guess mosquitto_passwd is necessary

**Reply by pfichtner** [reply]
Well, you can still run mosquitto_passwd on any other machine and transfer the file content but I guess you want to have a binary on the router that modifies the password file on the router directly.

**Comment by pfichtner** [comment, marked as answer]
Related to second question: Should be possible with #466

**Reply by pfichtner** [reply]
mosquitto_passwd available with PR #469

**Reply by pfichtner** [reply]
PR #469 was merged by @fda77 so mosquitto now can be built including `mosquitto_passwd`. Due to PR #466 the resulting file can be used using `password_file`

**Reply by tobit6795** [reply]
thanks for your efforts guys, i appreciate that

**Reply by pfichtner** [reply]
@tobit6795 please feel free to report any issues regarding `mosquitto_passwd` since it's completely untested at the moment (did not check more than that it compiles successfully). Also that mosquitto will work using the `password_file` option is untested for now. If it's working you could mark the question as answered. Thank you

**Reply by fda77** [reply]
Thx pfichtner! The "mark answer" is tricky on Github. Only "root" posts could be marked

**Reply by tobit6795** [reply]
of course - will do. i already built an image, but for some reasons my 6591 boxes wont flash over SSH anymore (invalid sector -1). ill try to give you feedback till end of the week

**Reply by tobit6795** [reply]
hi, sorry for the late response - i just flashed a 7590 and tested the passwd utility, works as expected, thanks again! :)

**Reply by tobit6795** [reply]
@pfichtner i finally started doing some configuration and it looks like there is the auth plugin missing

this is how it should look in configuration
```bash
# Authentication plugin
auth_plugin /usr/share/mosquitto/go-auth.so
```

this is what i found on my box

```
root@fritz-KB:/var/mod/etc/conf# find / -name "*mo
squitto*"
/etc/default.mosquitto
/etc/default.mosquitto/mosquitto.cfg
/etc/default.mosquitto/mosquitto_conf
/etc/default.mosquitto/mosquitto_extra.def
/etc/init.d/rc.mosquitto
find: /sys/kernel/slab/L2TP/IPv6: No such file or directory
find: /sys/kernel/slab/L2TP/IP: No such file or directory
/usr/bin/mosquitto
/usr/bin/mosquitto_passwd
/usr/bin/mosquitto_pub
/usr/bin/mosquitto_sub
/usr/lib/cgi-bin/mosquitto.cgi
/usr/lib/freetz/libmosquitto.so.1
/var/mod/usr/lib/cgi-bin/mosquitto.cgi
/var/mod/pkg/mosquitto
/var/mod/var/cache/menu/pkg/mosquitto
/var/mod/var/cache/menu/pkg/mosquitto.sub
/var/mod/etc/mosquitto.conf
/var/mod/etc/default.mosquitto
/var/mod/etc/init.d/rc.mosquitto
/var/mod/etc/conf/mosquitto.cfg
/var/tmp/flash/mosquitto
/var/tmp/flash/mosquitto/mosquitto.extra
```

How can we add the auth plugin?

**Comment by pfichtner** [comment]
What do you mean by _**the** auth plugin_? `mosquitto` does not need a  auth plugin for passwd-based authentication (and that's what `mosquitto_passwd` is for), see https://mosquitto.org/documentation/authentication-methods/

[go-auth](https://github.com/iegomez/mosquitto-go-auth) is one of a bunch of third-party mosquitto plugins which of course are not build just be enabling `mosquitto_passwd`. 

But what's the problem? Why did you ask for enabling `mosquitto_passwd`? 
`mosquitto_passwd` is for adding user's password hashes to a passwd file which can be used by mosquitto when 
```
allow_anonymous false
password_file /path/to/password/file
```
gets added to mosquitto's config file. 

`go-auth` is a plugin to enable mosquitto to check the password not just against a passwd file but to retrieve thoses hashes from other sources like databases and so on, so `mosquitto_passwd` and `go-auth` have nothing in common. 

So what's your need for `go-auth`? And if you really have the need for it you would have to build it separately because as I said: This is nor part of mosquitto but some third party module.

**Comment by tobit6795** [comment]
Well, my bad - should have read the documentation instead of trying to reuse my old configuration. Thanks for the explanation.
